features/mbedtls/platform/src/threading_alt.cpp

```diff
@@ -0,0 +1,79 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_THREADING_C)
+
+#include "mbedtls/threading.h"
+
+#if defined(MBEDTLS_THREADING_ALT)
+static void threading_mutex_init_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL )
+        return;
+
+    mutex->mutex = new rtos::Mutex();
+
+    mutex->is_valid = true;
+}
+
+static void threading_mutex_free_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid != true )
+        return;
+
+    delete mutex->mutex;
+
+    mutex->is_valid = false;
+}
+
+static int threading_mutex_lock_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid != true )
+        return( MBEDTLS_ERR_THREADING_BAD_INPUT_DATA );
+
+    if( mutex->mutex->lock() != osOK )
+        return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+
+    return( 0 );
+}
+
+static int threading_mutex_unlock_mbed( mbedtls_threading_mutex_t *mutex )
+{
+    if( mutex == NULL || mutex->is_valid != true )
+        return( MBEDTLS_ERR_THREADING_BAD_INPUT_DATA );
+
+    if( mutex->mutex->unlock() != osOK )
+        return( MBEDTLS_ERR_THREADING_MUTEX_ERROR );
+
+    return( 0 );
+}
+
+void mbedtls_threading_set_mbed( void )
```

Comment:
I think this function is confusingly named, and I'd prefer we didn't have it. Really the library should self-configure and this function causes life time issues, because now something has to call it during initialisation to allow the library to use the primitives. Currently the caller would be the mbed OS application developer - and that is unacceptable as it means we're adding a breaking API change that extends the mbed TLS library interface, but is also outside the library.

This is a library problem, which I'll raise separately and will need to be fixed elsewhere before we can merge this PR.

Back to the point - while we have to have the function, can we call it something like `mbedtls_rtx_threading_init()`?

`_rtx_` because we're using the RTX primitives, and `_init` because it's more natural than `_set_alt()`.